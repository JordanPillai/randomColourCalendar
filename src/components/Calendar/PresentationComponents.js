import styled, { css } from 'styled-components';

const padding = '40px';
const margin = '0.188rem';

export const DefaultCalendarDay = styled.button`
  background: none;
  color: inherit;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: 0;
  padding: ${padding};
`;

export const CalendarGridItem = styled.div`
  ${props =>
    props.offset &&
    css`
      grid-column-start: ${props.offset};
    `};
  border: 1px solid;
  font-size: 0.8125rem;
  margin-left: -1px;
  margin-bottom: -1px;
  & > * {
    padding: ${padding};
    display: block;
    width: 100%;
  }
  text-align: center;
`;

export const CalendarMonth = styled.div`
  &:not(:last-child) {
    margin-right: 1rem;
  }
  display: inline-block;
`;

export const CalendarGridHeading = CalendarGridItem.extend`
  color: black;
  opacity: 0.35;
  padding: ${padding};
`;

export const CalendarGrid = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr 1fr;
  text-align: center;
  border: 1px solid #c4c4c4;
`;

export const CalendarControl = styled.div`
  margin: ${props =>
    props.rightArrow ? '0 ' + margin + ' 0 auto' : '0 auto 0 ' + margin};
  padding: 0 ${padding};
  width: 32px;
  box-sizing: border-box;
  height: 18px;
`;

export const CalendarArrow = styled.button`
  background: none;
  font-weight: bold;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: 0;
  padding: 0;
`;

export const CalendarMonthHeading = styled.span`
  text-transform: uppercase;
  text-align: center;
  display: block;
  font-size: 0.875rem;
`;
