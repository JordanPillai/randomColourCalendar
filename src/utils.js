import subMinutes from 'date-fns/sub_minutes';
import eachDay from 'date-fns/each_day';
import isFirstDayOfMonth from 'date-fns/is_first_day_of_month';

export const adjustToUtc = date => {
  const offset = date.getTimezoneOffset();
  const adjustedDate = subMinutes(date, offset);
  return adjustedDate;
};

export const eachMonth = (startDate, endDate) => {
  return eachDay(startDate, endDate).filter(isFirstDayOfMonth);
};
