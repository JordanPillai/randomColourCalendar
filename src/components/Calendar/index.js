import React from 'react';
import styled from 'styled-components';

import startOfMonth from 'date-fns/start_of_month';
import getDayOfYear from 'date-fns/get_day_of_year';
import subMonths from 'date-fns/sub_months';
import eachDay from 'date-fns/each_day';
import addMonths from 'date-fns/add_months';
import endOfMonth from 'date-fns/end_of_month';
import format from 'date-fns/format';
import getDay from 'date-fns/get_day';
import getMonth from 'date-fns/get_month';

import {
  CalendarMonth,
  CalendarControl,
  CalendarArrow,
  CalendarMonthHeading,
  CalendarGrid,
  CalendarGridHeading,
  CalendarGridItem,
  DefaultCalendarDay
} from './PresentationComponents';

import { eachMonth } from '../../utils';

const abbreviatedDaysInWeek = [
  'Mon',
  'Tues',
  'Weds',
  'Thurs',
  'Fri',
  'Sat',
  'Sun'
];

const offsetForDays = [7, 1, 2, 3, 4, 5, 6];

const Flex = styled.div`
  display: flex;
  padding: 20px;
`;

export default function Calendar(props) {
  const {
    start = new Date(),
    noOfMonths = 1,
    onNext,
    onPrevious,
    renderDay,
    onDaySelect
  } = props;
  const startMonth = startOfMonth(start);
  const endMonth = addMonths(startMonth, noOfMonths - 1);
  const months = eachMonth(startMonth, endMonth);
  const daysInMonths = months.map(month => {
    const endDate = endOfMonth(month);
    const day = eachDay(month, endDate);
    return day;
  });
  return (
    <div>
      {daysInMonths.map((daysInMonth, index) => {
        return (
          <CalendarMonth key={'month-' + getMonth(daysInMonth[0])}>
            <Flex>
              <CalendarControl>
                {index === 0 ? (
                  <CalendarArrow
                    onClick={() =>
                      onPrevious && onPrevious(subMonths(start, 1))
                    }
                  >
                    Prev
                  </CalendarArrow>
                ) : null}
              </CalendarControl>
              <CalendarMonthHeading>
                {format(daysInMonth[0], 'MMM YYYY')}
              </CalendarMonthHeading>
              <CalendarControl rightArrow>
                {index === noOfMonths - 1 ? (
                  <CalendarArrow
                    onClick={() => onNext && onNext(addMonths(start, 1))}
                  >
                    Next
                  </CalendarArrow>
                ) : null}
              </CalendarControl>
            </Flex>
            <CalendarGrid>
              {abbreviatedDaysInWeek.map((dayName, i) => {
                return (
                  <CalendarGridHeading key={'day-name-' + i}>
                    {dayName}
                  </CalendarGridHeading>
                );
              })}
              {daysInMonth.map((day, index) => {
                return (
                  <CalendarGridItem
                    onClick={() => onDaySelect && onDaySelect(day)}
                    offset={
                      index === 0 ? offsetForDays[getDay(day)] : undefined
                    }
                    key={'day-' + getDayOfYear(day)}
                  >
                    {renderDay ? (
                      renderDay(format(day, 'D'), DefaultCalendarDay, day)
                    ) : (
                      <DefaultCalendarDay>
                        {format(day, 'D')}
                      </DefaultCalendarDay>
                    )}
                  </CalendarGridItem>
                );
              })}
            </CalendarGrid>
          </CalendarMonth>
        );
      })}
    </div>
  );
}
