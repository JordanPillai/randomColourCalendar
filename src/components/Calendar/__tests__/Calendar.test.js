import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import Calendar from '../index';
import { CalendarArrow } from '../PresentationComponents';
import { shallow } from 'enzyme';
import { adjustToUtc } from '../../../utils';
test('renders and matches snapshot', () => {
  const start = adjustToUtc(new Date('2018-04-01'));
  const wrapper = shallow(<Calendar start={start} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('onNext event', () => {
  const myMock = jest.fn();
  const start = new Date('2018-04-01');
  const wrapper = shallow(<Calendar onNext={myMock} start={start} />);

  // Next button
  wrapper
    .find(CalendarArrow)
    .at(1)
    .simulate('click');
  expect(toJson(wrapper)).toMatchSnapshot();
  expect(adjustToUtc(myMock.mock.calls[0][0]).getTime()).toBe(
    adjustToUtc(new Date('2018-05-01')).getTime()
  );
});

test('onPrevious event', () => {
  const myMock = jest.fn();
  const start = new Date('2018-03-01');
  const wrapper = shallow(<Calendar onPrevious={myMock} start={start} />);

  // Prev button
  wrapper
    .find(CalendarArrow)
    .at(0)
    .simulate('click');
  expect(toJson(wrapper)).toMatchSnapshot();
  expect(adjustToUtc(myMock.mock.calls[0][0]).getTime()).toBe(
    adjustToUtc(new Date('2018-02-01')).getTime()
  );
});
