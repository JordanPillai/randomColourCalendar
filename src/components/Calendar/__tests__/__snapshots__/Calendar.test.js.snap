// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onNext event 1`] = `
<div>
  <styled.div
    key="month-3"
  >
    <styled.div>
      <styled.div>
        <styled.button
          onClick={[Function]}
        >
          Prev
        </styled.button>
      </styled.div>
      <styled.span>
        Apr 2018
      </styled.span>
      <styled.div
        rightArrow={true}
      >
        <styled.button
          onClick={[Function]}
        >
          Next
        </styled.button>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div
        key="day-name-0"
      >
        Mon
      </styled.div>
      <styled.div
        key="day-name-1"
      >
        Tues
      </styled.div>
      <styled.div
        key="day-name-2"
      >
        Weds
      </styled.div>
      <styled.div
        key="day-name-3"
      >
        Thurs
      </styled.div>
      <styled.div
        key="day-name-4"
      >
        Fri
      </styled.div>
      <styled.div
        key="day-name-5"
      >
        Sat
      </styled.div>
      <styled.div
        key="day-name-6"
      >
        Sun
      </styled.div>
      <styled.div
        key="day-91"
        offset={7}
        onClick={[Function]}
      >
        <styled.button>
          1
        </styled.button>
      </styled.div>
      <styled.div
        key="day-92"
        onClick={[Function]}
      >
        <styled.button>
          2
        </styled.button>
      </styled.div>
      <styled.div
        key="day-93"
        onClick={[Function]}
      >
        <styled.button>
          3
        </styled.button>
      </styled.div>
      <styled.div
        key="day-94"
        onClick={[Function]}
      >
        <styled.button>
          4
        </styled.button>
      </styled.div>
      <styled.div
        key="day-95"
        onClick={[Function]}
      >
        <styled.button>
          5
        </styled.button>
      </styled.div>
      <styled.div
        key="day-96"
        onClick={[Function]}
      >
        <styled.button>
          6
        </styled.button>
      </styled.div>
      <styled.div
        key="day-97"
        onClick={[Function]}
      >
        <styled.button>
          7
        </styled.button>
      </styled.div>
      <styled.div
        key="day-98"
        onClick={[Function]}
      >
        <styled.button>
          8
        </styled.button>
      </styled.div>
      <styled.div
        key="day-99"
        onClick={[Function]}
      >
        <styled.button>
          9
        </styled.button>
      </styled.div>
      <styled.div
        key="day-100"
        onClick={[Function]}
      >
        <styled.button>
          10
        </styled.button>
      </styled.div>
      <styled.div
        key="day-101"
        onClick={[Function]}
      >
        <styled.button>
          11
        </styled.button>
      </styled.div>
      <styled.div
        key="day-102"
        onClick={[Function]}
      >
        <styled.button>
          12
        </styled.button>
      </styled.div>
      <styled.div
        key="day-103"
        onClick={[Function]}
      >
        <styled.button>
          13
        </styled.button>
      </styled.div>
      <styled.div
        key="day-104"
        onClick={[Function]}
      >
        <styled.button>
          14
        </styled.button>
      </styled.div>
      <styled.div
        key="day-105"
        onClick={[Function]}
      >
        <styled.button>
          15
        </styled.button>
      </styled.div>
      <styled.div
        key="day-106"
        onClick={[Function]}
      >
        <styled.button>
          16
        </styled.button>
      </styled.div>
      <styled.div
        key="day-107"
        onClick={[Function]}
      >
        <styled.button>
          17
        </styled.button>
      </styled.div>
      <styled.div
        key="day-108"
        onClick={[Function]}
      >
        <styled.button>
          18
        </styled.button>
      </styled.div>
      <styled.div
        key="day-109"
        onClick={[Function]}
      >
        <styled.button>
          19
        </styled.button>
      </styled.div>
      <styled.div
        key="day-110"
        onClick={[Function]}
      >
        <styled.button>
          20
        </styled.button>
      </styled.div>
      <styled.div
        key="day-111"
        onClick={[Function]}
      >
        <styled.button>
          21
        </styled.button>
      </styled.div>
      <styled.div
        key="day-112"
        onClick={[Function]}
      >
        <styled.button>
          22
        </styled.button>
      </styled.div>
      <styled.div
        key="day-113"
        onClick={[Function]}
      >
        <styled.button>
          23
        </styled.button>
      </styled.div>
      <styled.div
        key="day-114"
        onClick={[Function]}
      >
        <styled.button>
          24
        </styled.button>
      </styled.div>
      <styled.div
        key="day-115"
        onClick={[Function]}
      >
        <styled.button>
          25
        </styled.button>
      </styled.div>
      <styled.div
        key="day-116"
        onClick={[Function]}
      >
        <styled.button>
          26
        </styled.button>
      </styled.div>
      <styled.div
        key="day-117"
        onClick={[Function]}
      >
        <styled.button>
          27
        </styled.button>
      </styled.div>
      <styled.div
        key="day-118"
        onClick={[Function]}
      >
        <styled.button>
          28
        </styled.button>
      </styled.div>
      <styled.div
        key="day-119"
        onClick={[Function]}
      >
        <styled.button>
          29
        </styled.button>
      </styled.div>
      <styled.div
        key="day-120"
        onClick={[Function]}
      >
        <styled.button>
          30
        </styled.button>
      </styled.div>
    </styled.div>
  </styled.div>
</div>
`;

exports[`onPrevious event 1`] = `
<div>
  <styled.div
    key="month-2"
  >
    <styled.div>
      <styled.div>
        <styled.button
          onClick={[Function]}
        >
          Prev
        </styled.button>
      </styled.div>
      <styled.span>
        Mar 2018
      </styled.span>
      <styled.div
        rightArrow={true}
      >
        <styled.button
          onClick={[Function]}
        >
          Next
        </styled.button>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div
        key="day-name-0"
      >
        Mon
      </styled.div>
      <styled.div
        key="day-name-1"
      >
        Tues
      </styled.div>
      <styled.div
        key="day-name-2"
      >
        Weds
      </styled.div>
      <styled.div
        key="day-name-3"
      >
        Thurs
      </styled.div>
      <styled.div
        key="day-name-4"
      >
        Fri
      </styled.div>
      <styled.div
        key="day-name-5"
      >
        Sat
      </styled.div>
      <styled.div
        key="day-name-6"
      >
        Sun
      </styled.div>
      <styled.div
        key="day-60"
        offset={4}
        onClick={[Function]}
      >
        <styled.button>
          1
        </styled.button>
      </styled.div>
      <styled.div
        key="day-61"
        onClick={[Function]}
      >
        <styled.button>
          2
        </styled.button>
      </styled.div>
      <styled.div
        key="day-62"
        onClick={[Function]}
      >
        <styled.button>
          3
        </styled.button>
      </styled.div>
      <styled.div
        key="day-63"
        onClick={[Function]}
      >
        <styled.button>
          4
        </styled.button>
      </styled.div>
      <styled.div
        key="day-64"
        onClick={[Function]}
      >
        <styled.button>
          5
        </styled.button>
      </styled.div>
      <styled.div
        key="day-65"
        onClick={[Function]}
      >
        <styled.button>
          6
        </styled.button>
      </styled.div>
      <styled.div
        key="day-66"
        onClick={[Function]}
      >
        <styled.button>
          7
        </styled.button>
      </styled.div>
      <styled.div
        key="day-67"
        onClick={[Function]}
      >
        <styled.button>
          8
        </styled.button>
      </styled.div>
      <styled.div
        key="day-68"
        onClick={[Function]}
      >
        <styled.button>
          9
        </styled.button>
      </styled.div>
      <styled.div
        key="day-69"
        onClick={[Function]}
      >
        <styled.button>
          10
        </styled.button>
      </styled.div>
      <styled.div
        key="day-70"
        onClick={[Function]}
      >
        <styled.button>
          11
        </styled.button>
      </styled.div>
      <styled.div
        key="day-71"
        onClick={[Function]}
      >
        <styled.button>
          12
        </styled.button>
      </styled.div>
      <styled.div
        key="day-72"
        onClick={[Function]}
      >
        <styled.button>
          13
        </styled.button>
      </styled.div>
      <styled.div
        key="day-73"
        onClick={[Function]}
      >
        <styled.button>
          14
        </styled.button>
      </styled.div>
      <styled.div
        key="day-74"
        onClick={[Function]}
      >
        <styled.button>
          15
        </styled.button>
      </styled.div>
      <styled.div
        key="day-75"
        onClick={[Function]}
      >
        <styled.button>
          16
        </styled.button>
      </styled.div>
      <styled.div
        key="day-76"
        onClick={[Function]}
      >
        <styled.button>
          17
        </styled.button>
      </styled.div>
      <styled.div
        key="day-77"
        onClick={[Function]}
      >
        <styled.button>
          18
        </styled.button>
      </styled.div>
      <styled.div
        key="day-78"
        onClick={[Function]}
      >
        <styled.button>
          19
        </styled.button>
      </styled.div>
      <styled.div
        key="day-79"
        onClick={[Function]}
      >
        <styled.button>
          20
        </styled.button>
      </styled.div>
      <styled.div
        key="day-80"
        onClick={[Function]}
      >
        <styled.button>
          21
        </styled.button>
      </styled.div>
      <styled.div
        key="day-81"
        onClick={[Function]}
      >
        <styled.button>
          22
        </styled.button>
      </styled.div>
      <styled.div
        key="day-82"
        onClick={[Function]}
      >
        <styled.button>
          23
        </styled.button>
      </styled.div>
      <styled.div
        key="day-83"
        onClick={[Function]}
      >
        <styled.button>
          24
        </styled.button>
      </styled.div>
      <styled.div
        key="day-84"
        onClick={[Function]}
      >
        <styled.button>
          25
        </styled.button>
      </styled.div>
      <styled.div
        key="day-85"
        onClick={[Function]}
      >
        <styled.button>
          26
        </styled.button>
      </styled.div>
      <styled.div
        key="day-86"
        onClick={[Function]}
      >
        <styled.button>
          27
        </styled.button>
      </styled.div>
      <styled.div
        key="day-87"
        onClick={[Function]}
      >
        <styled.button>
          28
        </styled.button>
      </styled.div>
      <styled.div
        key="day-88"
        onClick={[Function]}
      >
        <styled.button>
          29
        </styled.button>
      </styled.div>
      <styled.div
        key="day-89"
        onClick={[Function]}
      >
        <styled.button>
          30
        </styled.button>
      </styled.div>
      <styled.div
        key="day-90"
        onClick={[Function]}
      >
        <styled.button>
          31
        </styled.button>
      </styled.div>
    </styled.div>
  </styled.div>
</div>
`;

exports[`renders and matches snapshot 1`] = `
<div>
  <styled.div
    key="month-3"
  >
    <styled.div>
      <styled.div>
        <styled.button
          onClick={[Function]}
        >
          Prev
        </styled.button>
      </styled.div>
      <styled.span>
        Apr 2018
      </styled.span>
      <styled.div
        rightArrow={true}
      >
        <styled.button
          onClick={[Function]}
        >
          Next
        </styled.button>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div
        key="day-name-0"
      >
        Mon
      </styled.div>
      <styled.div
        key="day-name-1"
      >
        Tues
      </styled.div>
      <styled.div
        key="day-name-2"
      >
        Weds
      </styled.div>
      <styled.div
        key="day-name-3"
      >
        Thurs
      </styled.div>
      <styled.div
        key="day-name-4"
      >
        Fri
      </styled.div>
      <styled.div
        key="day-name-5"
      >
        Sat
      </styled.div>
      <styled.div
        key="day-name-6"
      >
        Sun
      </styled.div>
      <styled.div
        key="day-91"
        offset={7}
        onClick={[Function]}
      >
        <styled.button>
          1
        </styled.button>
      </styled.div>
      <styled.div
        key="day-92"
        onClick={[Function]}
      >
        <styled.button>
          2
        </styled.button>
      </styled.div>
      <styled.div
        key="day-93"
        onClick={[Function]}
      >
        <styled.button>
          3
        </styled.button>
      </styled.div>
      <styled.div
        key="day-94"
        onClick={[Function]}
      >
        <styled.button>
          4
        </styled.button>
      </styled.div>
      <styled.div
        key="day-95"
        onClick={[Function]}
      >
        <styled.button>
          5
        </styled.button>
      </styled.div>
      <styled.div
        key="day-96"
        onClick={[Function]}
      >
        <styled.button>
          6
        </styled.button>
      </styled.div>
      <styled.div
        key="day-97"
        onClick={[Function]}
      >
        <styled.button>
          7
        </styled.button>
      </styled.div>
      <styled.div
        key="day-98"
        onClick={[Function]}
      >
        <styled.button>
          8
        </styled.button>
      </styled.div>
      <styled.div
        key="day-99"
        onClick={[Function]}
      >
        <styled.button>
          9
        </styled.button>
      </styled.div>
      <styled.div
        key="day-100"
        onClick={[Function]}
      >
        <styled.button>
          10
        </styled.button>
      </styled.div>
      <styled.div
        key="day-101"
        onClick={[Function]}
      >
        <styled.button>
          11
        </styled.button>
      </styled.div>
      <styled.div
        key="day-102"
        onClick={[Function]}
      >
        <styled.button>
          12
        </styled.button>
      </styled.div>
      <styled.div
        key="day-103"
        onClick={[Function]}
      >
        <styled.button>
          13
        </styled.button>
      </styled.div>
      <styled.div
        key="day-104"
        onClick={[Function]}
      >
        <styled.button>
          14
        </styled.button>
      </styled.div>
      <styled.div
        key="day-105"
        onClick={[Function]}
      >
        <styled.button>
          15
        </styled.button>
      </styled.div>
      <styled.div
        key="day-106"
        onClick={[Function]}
      >
        <styled.button>
          16
        </styled.button>
      </styled.div>
      <styled.div
        key="day-107"
        onClick={[Function]}
      >
        <styled.button>
          17
        </styled.button>
      </styled.div>
      <styled.div
        key="day-108"
        onClick={[Function]}
      >
        <styled.button>
          18
        </styled.button>
      </styled.div>
      <styled.div
        key="day-109"
        onClick={[Function]}
      >
        <styled.button>
          19
        </styled.button>
      </styled.div>
      <styled.div
        key="day-110"
        onClick={[Function]}
      >
        <styled.button>
          20
        </styled.button>
      </styled.div>
      <styled.div
        key="day-111"
        onClick={[Function]}
      >
        <styled.button>
          21
        </styled.button>
      </styled.div>
      <styled.div
        key="day-112"
        onClick={[Function]}
      >
        <styled.button>
          22
        </styled.button>
      </styled.div>
      <styled.div
        key="day-113"
        onClick={[Function]}
      >
        <styled.button>
          23
        </styled.button>
      </styled.div>
      <styled.div
        key="day-114"
        onClick={[Function]}
      >
        <styled.button>
          24
        </styled.button>
      </styled.div>
      <styled.div
        key="day-115"
        onClick={[Function]}
      >
        <styled.button>
          25
        </styled.button>
      </styled.div>
      <styled.div
        key="day-116"
        onClick={[Function]}
      >
        <styled.button>
          26
        </styled.button>
      </styled.div>
      <styled.div
        key="day-117"
        onClick={[Function]}
      >
        <styled.button>
          27
        </styled.button>
      </styled.div>
      <styled.div
        key="day-118"
        onClick={[Function]}
      >
        <styled.button>
          28
        </styled.button>
      </styled.div>
      <styled.div
        key="day-119"
        onClick={[Function]}
      >
        <styled.button>
          29
        </styled.button>
      </styled.div>
      <styled.div
        key="day-120"
        onClick={[Function]}
      >
        <styled.button>
          30
        </styled.button>
      </styled.div>
    </styled.div>
  </styled.div>
</div>
`;
